cgo/modecimal.h

```diff
@@ -0,0 +1,106 @@
+/* 
+ * Copyright 2021 Matrix Origin
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef _MODECIMAL_H_
+#define _MODECIMAL_H_
+
+#include "mo_impl.h"
+
+/*
+ * Mo's own impl of decimal64/128.  Only supports +,-,*,/,% and round.
+ * dec64 has precision of 17.
+ * dec128 has precision of 36.
+ *
+ * Binary format.
+ *      bit 1:     sign bit, 0 for neg, 1 for zero and pos.
+ *      bit 5/6:   number of integeral dec digits. 
+ *      intparts:  bits to store integral part, from above.
+ *                 intpart will be decimal digits, base 1000000000.
+ *                 The left most intpart could be short.
+ *      fracparts: bits to store fractional part, after intparts.
+ *
+ * If the number is negative (bit 1), the rest of the bits are
+ * reversed.
+ *
+ * Example:   123456789012345.6789 in decimal128 format.
+ *      1 bit    :      1       -- pos
+ *      6 bits   :      15      -- int part has 15 decimal digits
+ *      20 bits  :     123456   -- need 20 bits to store 6 digits
```

Comment:
To store 1 digit, need 4 bits, 2 digits 7, 3 digits, 10.    In general 10 bits is enough for 3 digits.   4, 7 bits for an extra 1 or 2 digits.

Reply:
I thought of using base 1000, instead of base 1000000000, but seem that only slow things down, we would rather use 1 int32 operation to add 9 digits, instead one int8 operation to add 3 digits.

Reply:
Why not use the less significant 57 bits to store int + frac parts as a whole? For this example, storing 1234567890123456789. Then add/sub/mul can use built-in operators. We only take care of the sign and the width/precision.